src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1909 @@
+
+        log_debug("Got version from job %" PRIu64 ": %s ", j->id, j->version);
+}
+
+static int job_on_exit(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        Job *j = ASSERT_PTR(userdata);
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *json = NULL;
+        int r;
+
+        assert(j);
+        assert(s);
+        assert(si);
+
+        if (IN_SET(j->type, JOB_UPDATE, JOB_VACUUM)) {
+                assert(j->target->busy);
+                j->target->busy = false;
+        }
+
+        if (si->si_code != CLD_EXITED) {
+                assert(IN_SET(si->si_code, CLD_KILLED, CLD_DUMPED));
+                sd_bus_error_setf(&error, SD_BUS_ERROR_FAILED,
+                                  "Job terminated abnormally with signal %s.",
+                                  signal_to_string(si->si_status));
+        } else if (si->si_status != EXIT_SUCCESS)
+                if (j->status_errno != 0)
+                        sd_bus_error_set_errno(&error, j->status_errno);
+                else
+                        sd_bus_error_setf(&error, SD_BUS_ERROR_FAILED,
+                                          "Job failed with exit code %i.", si->si_status);
+        else {
+                r = job_parse_child_output(TAKE_FD(j->stdout_fd), &json);
+                if (r < 0)
+                        sd_bus_error_set_errnof(&error, r, "Failed to parse JSON: %m");
+        }
+
+        /* Only send notification of exit if the job was actually detached */
+        if (j->detach_cb) {
+                r = sd_bus_emit_signal(
+                                j->manager->bus,
+                                "/org/freedesktop/sysupdate1",
+                                "org.freedesktop.sysupdate1.Manager",
+                                "JobRemoved",
+                                "toi",
+                                j->id,
+                                j->object_path,
+                                j->status_errno != 0 ? -j->status_errno : si->si_status);
+                if (r < 0)
+                        log_warning_errno(r, "Cannot emit JobRemoved message, ignoring: %m");
+        }
+
+        if (j->dbus_msg && j->complete_cb) {
+                if (sd_bus_error_is_set(&error)) {
+                        log_warning("Bus error occurred, ignoring callback for job: %s", error.message);
+                        sd_bus_reply_method_error(j->dbus_msg, &error);
+                } else {
+                        r = j->complete_cb(j->dbus_msg, j, json, &error);
+                        if (r < 0) {
+                                log_warning_errno(r, "Error during execution of job callback: %s", bus_error_message(&error, r));
+                                sd_bus_reply_method_errno(j->dbus_msg, r, &error);
+                        }
+                }
+        }
+
+        if (j->manager)
+                manager_check_idle(j->manager);
```

Comment:
> `job_free` is what removes the job from the manager's hash table

Yeah, this was weird from an ownership perspective. The manager owns the hash table, but jobs own the individual entries in the hash table? I was having trouble with this (double free IIRC), and this is the solution I found.

I don't think it's worth blocking the PR on this as the current solution works, if you think we can have something better we can do it in a follow-up?

Reply:
Current solution does not work. This is a bug.

As I said: your call to `manager_check_idle` doesn't do anything because the job is still in the manager's hash table. That call always does nothing because there's always at least one job in the table. Thus the manager is never idle when deciding if it should look for new targets. So it never looks for new targets. That functionality is just broken

Unlinking the object from the hash table in an object's destructor is a standard practice across the systemd code base. Even the hash table library is built to handle it

Reply:
I've moved the check after the `job_free` call.  It felt strange to have this called directly from `job_free`.  Removing the linkage between it and the manager in `job_free` makes sense, but making the manager update targets in there felt like too much coupling between the lifetime of jobs and the manager.